src/ray/util/pipe_logger.cc

```diff
@@ -152,37 +164,94 @@ std::shared_ptr<StreamDumper> CreateStreamDumper(ReadFunc read_func,
-  auto file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>(
-      fname, log_rotate_opt.rotation_max_size, log_rotate_opt.rotation_max_file_count);
+// Create a spdlog logger with all sinks specified by the given option.
+std::shared_ptr<spdlog::logger> CreateLogger(
+    const StreamRedirectionOption &log_redirect_opt) {
+  std::vector<spdlog::sink_ptr> logging_sinks;
+  spdlog::sink_ptr file_sink = nullptr;
+  if (log_redirect_opt.rotation_max_size != std::numeric_limits<size_t>::max()) {
+    file_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>(
+        log_redirect_opt.file_path,
+        log_redirect_opt.rotation_max_size,
+        log_redirect_opt.rotation_max_file_count);
+  } else {
+    file_sink =
+        std::make_shared<spdlog::sinks::basic_file_sink_st>(log_redirect_opt.file_path);
+  }
   file_sink->set_level(spdlog::level::info);
-
-  auto logger = std::make_shared<spdlog::logger>(fname, std::move(file_sink));
+  auto logger = std::make_shared<spdlog::logger>(
+      /*name=*/absl::StrFormat("pipe-logger-%s", log_redirect_opt.file_path),
+      std::move(file_sink));
+  logger->set_level(spdlog::level::info);
   logger->set_pattern("%v");  // Only message string is logged.
   return logger;
 }
 
+// Pipe streamer is only used in certain cases:
+// 1. Log roration is requested;
+// 2. Multiple sinks are involved.
+bool ShouldUsePipeStream(const StreamRedirectionOption &log_redirect_opt) {
+  const bool need_rotation =
+      log_redirect_opt.rotation_max_size != std::numeric_limits<size_t>::max();
+  return need_rotation || log_redirect_opt.tee_to_stdout ||
+         log_redirect_opt.tee_to_stderr;
+}
+
+#if defined(__APPLE__) || defined(__linux__)
+RedirectionFileHandle OpenFileForRedirection(const std::string &file_path) {
+  int fd = open(file_path.data(), O_WRONLY | O_CREAT, 0644);
+  RAY_CHECK_NE(fd, -1) << "Fails to open file " << file_path << " with failure reason "
+                       << strerror(errno);
+
+  auto flush_fn = [fd]() {
+    RAY_CHECK_EQ(fsync(fd), 0) << "Fails to flush data to disk because "
+                               << strerror(errno);
+  };
+  auto close_fn = [fd]() {
+    RAY_CHECK_EQ(fsync(fd), 0) << "Fails to flush data to disk because "
+                               << strerror(errno);
+    RAY_CHECK_EQ(close(fd), 0) << "Fails to close redirection file because "
+                               << strerror(errno);
+  };
+
+  return RedirectionFileHandle{fd, std::move(flush_fn), std::move(close_fn)};
+}
+#endif
+
 }  // namespace
 
-RotationFileHandle CreatePipeAndStreamOutput(const std::string &fname,
-                                             const LogRotationOption &log_rotate_opt,
-                                             std::function<void()> on_completion) {
 #if defined(__APPLE__) || defined(__linux__)
+RedirectionFileHandle CreateRedirectionFileHandle(
+    const StreamRedirectionOption &log_redirect_opt, const StdStreamFd &std_stream_fd) {
```

Comment:
No need to pass in StdStreamFd since it's only need for the `should_use_pipe_stream` case and we can just call `dup()` inside this function?

Reply:
updated